R/slide.R

```diff
@@ -318,36 +336,80 @@ epi_slide <- function(
+    # `.group_key`: we return early if `epi_slide`'s `.x` has 0 rows, and our
+    # loop over groups is the outer loop (>= 1 row into the group loop ensures
+    # we will have only 1-row `.group_key`s). Further, unlike `epix_slide`, we
+    # actually will be using `.group_data` rather than work with `.group_key` at
+    # all, in order to keep the pre-existing non-key columns. We will also try
+    # to work directly with `epi_df`s instead of listified tibbles; since we're
+    # not in as tight of a loop, the increased overhead hopefully won't matter.
+    # We'll need to use `bind_cols` rather than `c` to avoid losing
+    # `epi_df`ness.
+
+    res <- .data_group
+
+    if (is.null(.group_new_col_name)) {
+      if (inherits(slide_values, "data.frame")) {
+        # Sometimes slide_values can parrot back columns already in `res`; allow
+        # this, but balk if a column has the same name as one in `res` but a
+        # different value:
+        comp_nms <- names(slide_values)
+        overlaps_existing_names <- comp_nms %in% names(res)
+        for (comp_i in which(overlaps_existing_names)) {
+          if (!identical(slide_values[[comp_i]], res[[comp_nms[[comp_i]]]])) {
+            lines <- c(
+              cli::format_error(c(
+                "conflict detected between existing columns and slide computation output:",
+                "i" = "pre-existing columns: {syms(names(res))}",
+                "x" = "slide computation output included a column {syms(comp_nms[[comp_i]])} that didn't match the pre-existing value"
+              )),
+              capture.output(print(waldo::compare(res[[comp_nms[[comp_i]]]], slide_values[[comp_i]], x_arg = "existing", y_arg = "comp output"))),
+              cli::format_message(c("You likely want to rename or remove this column from your slide computation's output, or debug why it has a different value."))
+            )
+            rlang::abort(paste(collapse = "\n", lines),
+              class = "epiprocess__epi_slide_existing_vs_output_column_conflict"
+            )
+          }
         }
+        # Unpack into separate columns (without name prefix). If there are
+        # columns duplicating existing columns, de-dupe and order them as if they
+        # didn't exist in slide_values.
+        res <- bind_cols(res, slide_values[!overlaps_existing_names])
       } else {
-        # vector or packed data.frame-type column:
-        mutate(.data_group, !!.new_col_name := slide_values)
+        # Apply default name (to vector or packed data.frame-type column):
+        res[["slide_value"]] <- slide_values
+        # TODO check for bizarre conflicting `slide_value` existing col name.
+        # Either here or on entry to `epi_slide` (even if there we don't know
+        # whether vecs will be output). Or just turn this into a special case of
+        # the preceding branch and let the checking code there generate a
+        # complaint.
```

Comment:
See [this new validation](https://github.com/cmu-delphi/epiprocess/pull/519/files#diff-55607e70281a2d1f9409bef2d44b8479a6ad9fffb87168761db19d63da48e78aR185-R191).

Reply:
> We should probably just assert that "slide_value" not in names(.x) if .new_col_name is NULL?

If you mean at entry to epi[x]_slide, that's a little "imprecise" because the user comps might be outputting e.g. `tibble(a = 1, b = 2)` & expecting to be unpacked into 2 columns, so there's not actually a name conflict.
- Variant B: assert that "slide_value" is not in the epikey colnames of .x (still "imprecise" but nobody should be doing this anyway): user gets to do the above.  However, they also "get" to overwrite an existing slide_value value column.
- Alternative C: check around line 387, when we know there's an actual conflict.  Most "precise".  Might have significant overhead; need to benchmark whether it's stupid.
- Alternative D: complex restructuring to try to do all this name checking stuff only once..... too hard, not going to attempt now.

I'm not sure about A vs. B.  I'll test C first & if it's bad then we can decide A vs. B.

Reply:
> the user comps might be outputting e.g. tibble(a = 1, b = 2) & expecting to be unpacked into 2 columns, so there's not actually a name conflict.

So I get that the user computation returns a tibble for unpacking and the collision is handled below. That whole thing seems separate from allowing the user to specify names inside `.new_col_name` that collide with `names(.x)` at epi_slide start. Letting the user get away with colliding values in `.new_col_name` seems unnecessary to me - anyone expected unpackaed tibble outputs could just remove `.new_col_name` after the first error - and our validation is simplified.

Reply:
Sounds good.  To fit in with this approach, I'm thinking then we also not allow the names of unpacked tibble columns to conflict (except for the label/epikey columns which are allowed to parrot back where they came from).  & update docs to not say you can overwrite columns with a slide.

Reply:
I've done C in 477.  We can handle any performance implications later.  I don't think it's a major problem relative to the other tidyeval performance issues we noted in 477.